<?xml version="1.0" encoding="UTF-8"?>
<launch>
	<!-- ======================================================= arguments ==================================================== -->
	<arg name="ambient_pointcloud_topic" default="ambient_pointcloud" />
	<arg name="yaml_configuration_filters_filename" default="$(find dynamic_robot_localization)/yaml/configs/filters_2d.yaml" />
	<arg name="yaml_configuration_pose_estimation_filename" default="$(find dynamic_robot_localization)/yaml/configs/pose_estimation/initial_pose_estimation_2d.yaml" />
	<arg name="yaml_configuration_tracking_filename" default="$(find dynamic_robot_localization)/yaml/configs/pose_tracking/cluttered_environments_dynamic_2d.yaml" />
	<arg name="yaml_configuration_recovery_filename" default="$(find dynamic_robot_localization)/yaml/configs/pose_recovery/recovery_2d.yaml" />
	<arg name="reference_pointcloud_filename" default="" />
	<arg name="reference_pointcloud_preprocessed_save_filename" default="" />
	<arg name="reference_pointcloud_keypoints_filename" default="" />
	<arg name="reference_pointcloud_keypoints_save_filename" default="" />
	<arg name="reference_pointcloud_descriptors_filename" default="" />
	<arg name="reference_pointcloud_descriptors_save_filename" default="" />
	<arg name="publish_tf_map_odom" default="1" />
	<arg name="robot_initial_x" default="0.0" />
	<arg name="robot_initial_y" default="0.0" />
	<arg name="robot_initial_z" default="0.0" />
	<arg name="robot_initial_roll" default="0.0" />
	<arg name="robot_initial_pitch" default="0.0" />
	<arg name="robot_initial_yaw" default="0.0" />
	<arg name="use_laser_assembler" default="1" />
	<arg name="laser_scan_topics" default="/tilt_scan+/base_scan" />
	<arg name="number_of_scans_to_assemble_per_cloud" default="2" />
	<arg name="timeout_for_cloud_assembly" default="0.5" />
	<arg name="use_dynamic_map_update" default="0" />
	<arg name="octomap_file" default="" />


	<!-- ==================================================== localization system ============================================= -->
	<!-- <node pkg="dynamic_robot_localization" type="drl_localization_node" name="drl_localization_node" ns="dynamic_robot_localization" output="screen" launch-prefix="gdbserver localhost:1337" > -->
	<node pkg="dynamic_robot_localization" type="drl_localization_node" name="drl_localization_node" ns="dynamic_robot_localization" output="screen">
		<rosparam command="load" file="$(arg yaml_configuration_filters_filename)" subst_value="true" />
		<rosparam command="load" file="$(arg yaml_configuration_pose_estimation_filename)" subst_value="true" />
		<rosparam command="load" file="$(arg yaml_configuration_tracking_filename)" subst_value="true" />
		<rosparam command="load" file="$(arg yaml_configuration_recovery_filename)" subst_value="true" />
		<param name="subscribe_topic_names/ambient_pointcloud_topic" type="str" value="$(arg ambient_pointcloud_topic)" />
		<param name="reference_pointclouds/reference_pointcloud_filename" type="str" value="$(arg reference_pointcloud_filename)" />
		<param name="reference_pointclouds/reference_pointcloud_preprocessed_save_filename" type="str" value="$(arg reference_pointcloud_preprocessed_save_filename)" />
		<param name="reference_pointclouds/reference_pointcloud_keypoints_filename" type="str" value="$(arg reference_pointcloud_keypoints_filename)" />
		<param name="reference_pointclouds/reference_pointcloud_keypoints_save_filename" type="str" value="$(arg reference_pointcloud_keypoints_save_filename)" />
		<param name="cloud_matchers/feature_matchers/reference_pointcloud_descriptors_filename" type="str" value="$(arg reference_pointcloud_descriptors_filename)" />
		<param name="cloud_matchers/feature_matchers/reference_pointcloud_descriptors_save_filename" type="str" value="$(arg reference_pointcloud_descriptors_save_filename)" />
	</node>


	<!-- map to odom publisher -->
	<include file="$(find dynamic_robot_localization)/launch/pose_to_tf_publisher.launch" if="$(arg publish_tf_map_odom)" >
		<arg name="initial_x" default="$(arg robot_initial_x)" />
		<arg name="initial_y" default="$(arg robot_initial_y)" />
		<arg name="initial_z" default="$(arg robot_initial_z)" />
		<arg name="initial_roll" default="$(arg robot_initial_roll)" />
		<arg name="initial_pitch" default="$(arg robot_initial_pitch)" />
		<arg name="initial_yaw" default="$(arg robot_initial_yaw)" />
	</include>


	<!-- laser assembler -->
	<include file="$(find laserscan_to_pointcloud)/launch/laserscan_to_pointcloud_assembler.launch" ns="dynamic_robot_localization" if="$(arg use_laser_assembler)" >
		<arg name="laser_scan_topics" default="$(arg laser_scan_topics)" />
		<arg name="number_of_scans_to_assemble_per_cloud" default="$(arg number_of_scans_to_assemble_per_cloud)" />
		<arg name="timeout_for_cloud_assembly" default="$(arg timeout_for_cloud_assembly)" />
	</include>
	

	<!-- OctoMap for dynamic map update -->
	<include file="$(find dynamic_robot_localization)/launch/octo_map.launch" if="$(arg use_dynamic_map_update)" >
		<arg name="octomap_file" default="$(arg octomap_file)" />
	</include>
</launch>
 